lib/src/solver/package_lister.dart

```diff
@@ -0,0 +1,227 @@
+class PackageLister {
+  /// The package that is being listed.
+  final PackageRef _ref;
+
+  /// The source from which [_ref] comes.
+  final BoundSource _source;
+
+  /// A map from dependency names to constraints indicating which versions of
+  /// [_ref] have already had their dependencies on the given versions returned
+  /// by [incompatibilitiesFor].
+  ///
+  /// This allows us to avoid returning the same incompatibilities multiple
+  /// times.
+  final _alreadyListedDependencies = <String, VersionConstraint>{};
+
+  /// A constraint indicating which versions of [_ref] have already been listed
+  /// as not matching the current SDK version by [incompatibilitiesFor].
+  ///
+  /// This allows us to avoid returning the same incompatibilities multiple
+  /// times.
+  var _alreadyListedSdks = VersionConstraint.empty;
+
+  /// The versions of [_ref] that have been downloaded and cached, or `null` if
+  /// they haven't been downloaded yet.
+  List<PackageId> get cachedVersions => _versionsCache?.result?.asValue?.value;
+
+  /// All versions of the package, sorted by [Version.compareTo].
+  Future<List<PackageId>> get _versions {
+    _versionsCache ??=
+        new ResultFuture(_source.getVersions(_ref).then((versions) {
+      versions.sort((id1, id2) => id1.version.compareTo(id2.version));
+      return versions;
+    }));
+    return _versionsCache;
+  }
+
+  ResultFuture<List<PackageId>> _versionsCache;
+
+  PackageLister(SystemCache cache, this._ref)
+      : _source = cache.source(_ref.source);
+
+  /// Returns the number of versions of this package that match [constraint].
+  Future<int> countVersions(VersionConstraint constraint) async =>
+      (await _versions).where((id) => constraint.allows(id.version)).length;
+
+  /// Returns the latest version of this package that matches [constraint], or
+  /// `null` if no versions match.
+  Future<PackageId> latestVersion(VersionConstraint constraint) async {
+    var versions = await _versions;
+    var min = constraint is VersionRange ? constraint.min : null;
+
+    // Return the most preferable version that matches [constraint]: the latest
+    // non-prerelease version if one exists, or the latest prerelease version
+    // otherwise.
+    PackageId bestPrerelease;
+    for (var id in versions.reversed) {
+      // If [constraint] has a minimum, we can bail early once we're past it.
+      if (min != null && id.version < min) break;
+
+      if (!constraint.allows(id.version)) continue;
+      if (!id.version.isPreRelease) return id;
+      bestPrerelease ??= id;
+    }
+    return bestPrerelease;
+  }
+
+  /// Returns incompatibilities that encapsulate [id]'s dependencies, or that
+  /// indicate that it can't be safely selected.
+  ///
+  /// If multiple subsequent versions of this package have the same
+  /// dependencies, this will return incompatibilities that reflect that. It
+  /// won't return incompatibilities that have already been returned by a
```

Comment:
The method name reads as stateless to me, so it was surprising that it produces different results on each call. Maybe `nextIncompatibilitiesFor()`?

Reply:
It's in sort of a gray area. It's not strictly stateless in the sense that multiple calls with the same input won't produce the same output, but it's not exactly stateful in the sense that different calls don't affect the *semantics* of what's being returned. The caller doesn't get any *new* information from calling this again with the same id (which is why I don't love "next" as a descriptor), and the fact that the old information isn't repeated is kind of an implementation detail.